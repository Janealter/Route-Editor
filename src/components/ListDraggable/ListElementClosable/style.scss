.list-element {
  transition: border-color .10s ease-in-out;
  &.list-element_highlighted {
    border: 1px solid #b8daff;
  }
  .list-element__inner {
    display: flex;
    flex-direction: row;
    justify-content: space-between;
    align-items: center;
    .list-element__button {
      margin-left: 30px;
      padding-top: 0.2rem;
      padding-bottom: 0.2rem;
    }
  }
}