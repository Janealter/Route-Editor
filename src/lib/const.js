export const KeyCode = {
  ENTER: 13,
  ESC: 27
};