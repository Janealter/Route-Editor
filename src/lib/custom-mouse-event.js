export let getCustomMouseEvent = function(type, sx = 1, sy = 1, cx = 1, cy = 1) {
  let evt;
  const e = {
    bubbles: true,
    cancelable: (type !== "mousemove"),
    view: window,
    detail: 0,
    screenX: sx,
    screenY: sy,
    clientX: cx,
    clientY: cy,
    ctrlKey: false,
    altKey: false,
    shiftKey: false,
    metaKey: false,
    button: 0,
    relatedTarget: undefined
  };
  if (typeof  document.createEvent === "function") {
    evt = document.createEvent("MouseEvents");
    evt.initMouseEvent(type,
      e.bubbles, e.cancelable, e.view, e.detail,
      e.screenX, e.screenY, e.clientX, e.clientY,
      e.ctrlKey, e.altKey, e.shiftKey, e.metaKey,
      e.button, document.body.parentNode);
  } else if (document.createEventObject) {
    evt = document.createEventObject();
    for (prop in e) {
      evt[prop] = e[prop];
    }
    evt.button = {0: 1, 1: 4, 2: 2}[evt.button] || evt.button;
  }
  return evt;
}

export let dispatchEvent = function(el, evt) {
  if (el.dispatchEvent) {
    el.dispatchEvent(evt);
  } else if (el.fireEvent) {
    el.fireEvent('on' + type, evt);
  }
  return evt;
}

export let simulateMouseMove = function (startCoord, direction, distance) {
  for (let i = 1; i < distance; i++) {
    let mouseMove = null;
    switch (direction) {
      case 'up':
        mouseMove = getCustomMouseEvent("mousemove", 1, 1, 1, startCoord - i);
        break;
      case 'down':
        mouseMove = getCustomMouseEvent("mousemove", 1, 1, 1, startCoord + i);
        break;
      case 'left':
        mouseMove = getCustomMouseEvent("mousemove", 1, 1, startCoord - i, 1);
        break;
      case 'right':
        mouseMove = getCustomMouseEvent("mousemove", 1, 1, startCoord + i, 1);
        break;
    }
    dispatchEvent(window.document.documentElement, mouseMove);
  }
};
