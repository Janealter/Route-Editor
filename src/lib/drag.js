import ElementParameters from "./element-parameters";

export const startDrag = function (evt, moveEvtCB, upEvtCB, CBArgs, limitXMin, limitXMax, limitYMin, limitYMax) {
  evt.preventDefault();
  const elementParameters = new ElementParameters(evt.currentTarget);

  /*console.log(`MouseDown ScreenX: ${evt.screenX}`);
  console.log(`MouseDown ScreenY: ${evt.screenY}`);
  console.log(`MouseDown ClientX: ${evt.clientX}`);
  console.log(`MouseDown ClientY: ${evt.clientY}`);
  console.log('---------------------------------');*/

  // Element on top
  elementParameters.zIndex = '99';

  let BodySize = {
    WIDTH: parseInt(window.getComputedStyle(document.body).width, 10),
    HEIGHT: parseInt(window.getComputedStyle(document.body).height, 10)
  };

  limitXMin = limitXMin || 0;
  limitXMax = limitXMax || BodySize.WIDTH;
  limitYMin = limitYMin || 0;
  limitYMax = limitYMax || BodySize.HEIGHT;

  // Стартовые координаты курсора
  let mouseStartCoords = {
    x: evt.clientX,
    y: evt.clientY
  };

  // Смещение курсора относительно элемента
  let shift = {
    x: mouseStartCoords.x - elementParameters.x,
    y: mouseStartCoords.y - elementParameters.y
  };

  // Обработчик события при перемещении мыши
  let onElementMousemove = function (moveEvt) {
    /*console.log(`MouseMove ScreenX: ${moveEvt.screenX}`);
    console.log(`MouseMove ScreenY: ${moveEvt.screenY}`);
    console.log(`MouseMove ClientX: ${moveEvt.clientX}`);
    console.log(`MouseMove ClientY: ${moveEvt.clientY}`);
    console.log('---------------------------------');*/
    moveEvt.preventDefault();
    // Текущие координаты курсора
    let mouseCoords = {
      x: moveEvt.clientX,
      y: moveEvt.clientY
    };

    // Текущие координаты элемента
    let elementCoords = {
      x: mouseCoords.x - shift.x,
      y: mouseCoords.y - shift.y
    };

    if (elementCoords.x >= limitXMin && elementCoords.x + elementParameters.width <= limitXMax) {
      elementParameters.x = elementCoords.x;
    }
    if (elementCoords.y >= limitYMin && elementCoords.y + elementParameters.height <= limitYMax) {
      elementParameters.y = elementCoords.y;
    }
    if (typeof moveEvtCB === 'function') {
      moveEvtCB(elementParameters, ...CBArgs);
    }
  };

  // Обработчик события при отпускании кнопки мыши
  let onElementMouseup = function (upEvt) {
    upEvt.preventDefault();
    document.removeEventListener('mousemove', onElementMousemove);
    document.removeEventListener('mouseup', onElementMouseup);
    if (typeof upEvtCB === 'function') {
      upEvtCB(elementParameters, ...CBArgs);
    }
  };

  document.addEventListener('mousemove', onElementMousemove);
  document.addEventListener('mouseup', onElementMouseup);
};

// ID передвигаемых элементов должны состоять из любых 3-х символов и числа после них
// Число должно указывать на индекс элемента в массиве nodes
export const startDragListElement = function(evt, nodes, mouseUpCB) {
  const draggableElementParameters = new ElementParameters(evt.currentTarget);
  const draggableIndex = parseInt(draggableElementParameters.object.id.substring(3), 10);

  const dragInfo = {
    draggableElementStartIndex: draggableIndex,
    draggableElementIndex: draggableIndex,
    draggableElementPreviousY: draggableElementParameters.y,
    elementsRelativeY: getRelativeY(nodes, draggableIndex),
    nodes: nodes
  };

  const limitXMin = 0;
  const limitXMax = draggableElementParameters.width;
  const limitYMin = dragInfo.elementsRelativeY[0];
  const limitYMax = dragInfo.elementsRelativeY[dragInfo.elementsRelativeY.length - 1] +
    new ElementParameters(nodes[nodes.length - 1]).height;

  startDrag(evt,
    moveAdjacentListElements,
    mouseUpCB || null,
    [dragInfo], limitXMin ,limitXMax, limitYMin, limitYMax);
};

export const moveAdjacentListElements = function(draggableElementParameters, dragInfo) {
  // При движении вниз
  if (dragInfo.draggableElementIndex < dragInfo.elementsRelativeY.length - 1 &&
    draggableElementParameters.y > dragInfo.draggableElementPreviousY)
  {
    const nextElementParameters = new ElementParameters(dragInfo.nodes[dragInfo.draggableElementIndex + 1]);
    if (draggableElementParameters.y + draggableElementParameters.height >
      dragInfo.elementsRelativeY[dragInfo.draggableElementIndex + 1] + nextElementParameters.height / 2)
    {
      nextElementParameters.y -= draggableElementParameters.height - draggableElementParameters.border;
      dragInfo.nodes[dragInfo.draggableElementIndex] = dragInfo.nodes[dragInfo.draggableElementIndex + 1];
      dragInfo.nodes[dragInfo.draggableElementIndex + 1] = draggableElementParameters.object;
      dragInfo.elementsRelativeY[dragInfo.draggableElementIndex] =
        dragInfo.elementsRelativeY[dragInfo.draggableElementIndex + 1] - draggableElementParameters.height - draggableElementParameters.border;
      dragInfo.draggableElementIndex++;
    }
  }
  // При движении вверх
  else if (dragInfo.draggableElementIndex > 0 &&
    draggableElementParameters.y < dragInfo.draggableElementPreviousY)
  {
    const previousElementParameters = new ElementParameters(dragInfo.nodes[dragInfo.draggableElementIndex - 1]);
    if (draggableElementParameters.y <
      dragInfo.elementsRelativeY[dragInfo.draggableElementIndex - 1] + previousElementParameters.height / 2)
    {
      previousElementParameters.y += draggableElementParameters.height - draggableElementParameters.border;
      dragInfo.nodes[dragInfo.draggableElementIndex] = dragInfo.nodes[dragInfo.draggableElementIndex - 1];
      dragInfo.nodes[dragInfo.draggableElementIndex - 1] = draggableElementParameters.object;
      dragInfo.elementsRelativeY[dragInfo.draggableElementIndex] =
        dragInfo.elementsRelativeY[dragInfo.draggableElementIndex - 1] + draggableElementParameters.height - draggableElementParameters.border;
      dragInfo.draggableElementIndex--;
    }
  }
  dragInfo.draggableElementPreviousY = draggableElementParameters.y;
};

export const clearStyleAfterDragging = function(nodes) {
  nodes.forEach(element => {
    element.style.zIndex = '';
    element.style.top = '';
    element.style.left = '';
    element.style.borderRadius = '';
    element.style.height = '';
  });
};

// Возвращает список координат Y у массива DOM-элементов относительно элемента с индексом startIndex
export const getRelativeY = function (elementNodes, startIndex) {
  let relativeY = [];
  relativeY[startIndex] = 0;
  if (startIndex > 0) {
    let totalHeight = 0;
    for (let i = startIndex - 1; i >= 0; i--) {
      let elementParameters = new ElementParameters(elementNodes[i]);
      relativeY[i] = totalHeight - elementParameters.height + elementParameters.border;
      totalHeight += -elementParameters.height + elementParameters.border;
    }
  }
  if (startIndex < elementNodes.length - 1) {
    let totalHeight = 0;
    for (let i = startIndex + 1; i < elementNodes.length; i++) {
      let previousElementParameters = new ElementParameters(elementNodes[i - 1]);
      relativeY[i] = totalHeight + previousElementParameters.height - previousElementParameters.border;
      totalHeight += previousElementParameters.height - previousElementParameters.border;
    }
  }
  return relativeY;
};
