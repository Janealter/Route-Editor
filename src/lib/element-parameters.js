class ElementParameters {
  constructor(element) {
    this.object = element;
  }

  get x() {
    return parseFloat(window.getComputedStyle(this.object).left);
  }

  set x(v) {
    this.object.style.left = v + 'px';
  }

  get y() {
    return parseFloat(window.getComputedStyle(this.object).top);
  }

  set y(v) {
    this.object.style.top = v + 'px';
  }

  get width() {
    return parseFloat(window.getComputedStyle(this.object).width);
  }

  set width(v) {
    this.object.style.width = v + 'px';
  }

  get height() {
    return parseFloat(window.getComputedStyle(this.object).height);
  }

  set height(v) {
    this.object.style.height = v + 'px';
  }

  get border() {
    const border = window.getComputedStyle(this.object).border;
    return border !== '' ? parseFloat(border) : 0;
  }

  set zIndex(v) {
    this.object.style.zIndex = v;
  }

  get zIndex() {
    return parseInt(this.object.style.zIndex, 10);
  }

}

export default ElementParameters;
