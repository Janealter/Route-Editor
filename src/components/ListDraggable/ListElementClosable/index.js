import React, { Component } from 'react';
import PropTypes from 'prop-types';
import 'bootstrap/dist/css/bootstrap.css';
import './style.scss';

class ListElementClosable extends Component {
  constructor(props) {
    super(props);
    this.handleMouseDown = this.handleMouseDown.bind(this);
    this.handleMouseUp = this.handleMouseUp.bind(this);
    this.state = {
      highlighted: false
    };
  }

  handleMouseDown(evt) {
    if (this.props.isHighlightable) {
      // Добавляем listener отпускания кнопки мыши на все окно, чтобы подсветка элемента выключалась
      // при отпускании кнопки в любом месте, а не только над элементом
      window.addEventListener('mouseup', this.handleMouseUp);
      this.setState({ highlighted: true });
    }
    if(typeof this.props.onMouseDown === 'function') {
      this.props.onMouseDown(evt);
    }
  }

  handleMouseUp() {
    window.removeEventListener('mouseup', this.handleMouseUp);
    this.setState({ highlighted: false });
  }

  render() {
    return (
      <li ref={this.props.onElementNodeCreate}
          id={this.props.id}
          className={`list-element list-group-item${this.state.highlighted ? ' list-element_highlighted' : ''}`}
          onMouseDown={this.handleMouseDown}
      >
        <div className="list-element__inner">
          <span className="list-element__text">{this.props.text}</span>
          {this.props.isClosable ? <button id={this.props.id} className="list-element__button btn btn-danger btn-sm float-right" onClick={this.props.onCloseButtonClick}>x</button> : ''}
        </div>
      </li>
    );
  }
}

ListElementClosable.propTypes = {
  id: PropTypes.string,
  text: PropTypes.string,
  isClosable: PropTypes.bool,
  isHighlightable: PropTypes.bool,
  onMouseDown: PropTypes.func,
  onCloseButtonClick: PropTypes.func,
  onElementNodeCreate: PropTypes.func
};

export default ListElementClosable;
