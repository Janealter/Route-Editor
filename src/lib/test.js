import {startDrag, startDragListElement, clearStyleAfterDragging, getRelativeY} from './drag';
import ElementParameters from './element-parameters';
import { dispatchEvent, getCustomMouseEvent, simulateMouseMove } from "./custom-mouse-event";

const createListAndGetListElementNodes = function() {
  const list = document.createElement('ul');
  const listElements = document.createDocumentFragment();

  for (let i = 0; i < 5; i++) {
    const listElement = document.createElement('li');
    listElements.appendChild(listElement);
  }
  list.appendChild(listElements);
  return [].map.call(list.querySelectorAll('li'), listElement => listElement);
};

document.body.style.height = '600px';
document.body.style.width = '800px';

describe('startDrag', function () {
  const element = document.createElement('div');

  it('is work fine', () => {
    element.style.top = '0px';
    element.style.left = '0px';
    element.style.height = '53.345px';
    element.style.width = '118.25px';

    element.addEventListener('mousedown', evt => {
      startDrag(evt);
    });

    dispatchEvent(element, getCustomMouseEvent("mousedown"));
    simulateMouseMove(1, 'down', 110);
    dispatchEvent(window.document.documentElement, getCustomMouseEvent("mouseup"));
  });

  it('the snapshot and the style are the same', () => {
    expect(JSON.stringify(element.style)).toMatchSnapshot();
  });

});

describe('startDragListElement', function () {
  const listElementNodes = createListAndGetListElementNodes();
  const draggedElement = listElementNodes[1];

  it('is work fine', () => {
    // Указываем высоту элементов, координаты и бордеры вручную,
    // так как getComputedStyle в среде JSDOM возвращает пустые строки вместо значений
    const elementHeights = [
      '85.75px',
      '167.35px',
      '15.375px',
      '17.65px',
      '35.55px',
    ];
    const draggedElementParameters = new ElementParameters(draggedElement);
    // Эталонное состояние dragInfo перед передвижением элемента
    const startDragInfo = {
      draggableElementStartIndex: 1,
      draggableElementIndex: 1,
      draggableElementPreviousY: 0,
      elementsRelativeY: [ -84.75, 0, 166.35, 180.725, 197.375 ],
      nodes: listElementNodes
    };
    // Эталонное состояние dragInfo после передвижения элемента
    const finalDragInfo = {
      draggableElementStartIndex: 1,
      draggableElementIndex: 3,
      draggableElementPreviousY: 29,
      elementsRelativeY: [ -84.75, -2, 12.375, 180.725, 197.375 ],
      nodes: listElementNodes
    };
    const onElementMouseDownSpy = jest.fn();//(ep, di) => console.log(di);
    const onElementMouseUpSpy = jest.fn();
    elementHeights.forEach((height, index) => {
      listElementNodes[index].id = `le-${index}`;
      listElementNodes[index].style.top = '0px';
      listElementNodes[index].style.left = '0px';
      listElementNodes[index].style.height = height;
      listElementNodes[index].style.width = `${height * 3}px`;
      listElementNodes[index].style.border = '1px';
    });

    draggedElement.addEventListener('mousedown', evt => {
      startDragListElement(evt, listElementNodes, onElementMouseDownSpy, onElementMouseUpSpy);
    });

    dispatchEvent(listElementNodes[1], getCustomMouseEvent("mousedown"));
    expect(onElementMouseDownSpy).toBeCalledWith(draggedElementParameters, startDragInfo);
    simulateMouseMove(1, 'down', 30);
    simulateMouseMove(191, 'up', 10);
    dispatchEvent(window.document.documentElement, getCustomMouseEvent("mouseup"));
    expect(onElementMouseUpSpy).toBeCalledWith(draggedElementParameters, finalDragInfo);
  });

  it('the snapshot and the style are the same', () => {
    expect(JSON.stringify([].map.call(listElementNodes, node => node.style))).toMatchSnapshot();
  });
});

describe('clearStyleAfterDragging', function () {
  const element = document.createElement('div');

  it('is work fine', () => {
    element.style.top = '0px';
    element.style.left = '0px';
    element.style.height = '53.345px';
    element.style.width = '118.25px';
    element.style.zIndex = '5';
    element.style.borderRadius = '2px';

    clearStyleAfterDragging([element]);
  });

  it('the snapshot and the style are the same', () => {
    expect(JSON.stringify(element.style)).toMatchSnapshot();
  });
});
