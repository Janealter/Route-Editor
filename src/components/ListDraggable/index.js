import React, { Component } from 'react';
import PropTypes from 'prop-types';
import ListElementClosable from './ListElementClosable';
import { clearStyleAfterDragging, startDragListElement } from "../../lib/drag";

class ListDraggable extends Component {
  constructor(props) {
    super(props);
    this.handleElementMouseDown = this.handleElementMouseDown.bind(this);
    this.handleElementMouseUp = this.handleElementMouseUp.bind(this);
    this.handleElementDelete = this.handleElementDelete.bind(this);
    this.handleElementNodeCreate = this.handleElementNodeCreate.bind(this);
  }

  handleElementMouseDown(evt) {
    // Не выполнять передвижение, если была нажата кнопка или выделен текст
    // || evt.target.localName === 'span'
    if (evt.target.localName === 'button') {
      return false;
    }

    // В начале передвижения убираем округление углов, так красивее
    this.elementNodes.forEach(elementNode => {
      elementNode.style.borderRadius = '0';
    });

    startDragListElement(evt, this.elementNodes, this.handleElementMouseUp);
  }

  handleElementMouseUp(draggableElementParameters, dragInfo) {
    // Очищаем стиль элемента, иначе у элемента останутся аттрибуты, установленые во время передвижения
    // и начальное позиционирование после обновления данных будет неправильным
    clearStyleAfterDragging(this.elementNodes);

    if (typeof this.props.onElementSort === 'function') {
      this.props.onElementSort(dragInfo.draggableElementStartIndex, dragInfo.draggableElementIndex);
    }
  }

  handleElementDelete(evt) {
    evt.preventDefault();
    // Удаляем элемент из DOM
    const id = parseInt(evt.currentTarget.id.substring(3), 10);
    this.elementNodes.splice(id, 1);
    if (typeof this.props.onElementDelete === 'function') {
      const key = this.props.elements[id].key;
      this.props.onElementDelete(key);
    }
  }

  handleElementNodeCreate(listElementNode) {
    if (listElementNode !== null) {
      this.elementNodes.push(listElementNode);
    }
  }

  elementNodes = [];

  render() {
    const ListElementTemplate = this.props.listElementTemplate || ListElementClosable;
    const elements = this.props.elements.map((element, index) =>
      <ListElementTemplate onElementNodeCreate={this.handleElementNodeCreate}
                           id={'le-' + index} key={element.key} text={element.text}
                           onMouseDown={this.handleElementMouseDown}
                           onCloseButtonClick={this.handleElementDelete}
                           isClosable={element.isRemovable}
                           isHighlightable
      />);
    return (
      <ul className="list-group">
        {elements}
      </ul>
    );
  }
}

ListDraggable.propTypes = {
  elements: PropTypes.arrayOf(PropTypes.object).isRequired,
  listElementTemplate: PropTypes.func,
  onElementDelete: PropTypes.func,
  onElementSort: PropTypes.func
};

export default ListDraggable;
