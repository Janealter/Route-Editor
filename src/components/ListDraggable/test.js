import React from 'react';
import { mount } from 'enzyme';
import ListElementClosable from "../ElementClosable";
import ListDraggable from "./index";
import { dispatchEvent, getCustomMouseEvent, simulateMouseMove } from "../../lib/custom-mouse-event";
import toJson from "enzyme-to-json";

describe("ListDraggable", function() {
  const elements = [
    {
      key: 1,
      text: 'element1',
      isRemovable: true
    },
    {
      key: 2,
      text: 'element2',
      isRemovable: true
    },
    {
      key: 3,
      text: 'element3',
      isRemovable: true
    },
  ];
  const onElementDeleteSpy = jest.fn();
  const onElementSortSpy = jest.fn();
  const renderedComponent = mount(
    <ListDraggable elements={elements}
                   listElementTemplate={ListElementClosable}
                   onElementDelete={onElementDeleteSpy}
                   onElementSort={onElementSortSpy}
  />);

  it('renders self and inner components', () => {
    expect(renderedComponent.instance()).toBeInstanceOf(ListDraggable);
    expect(renderedComponent.props().elements).toEqual(elements);
    expect(renderedComponent.props().listElementTemplate).toEqual(ListElementClosable);
    expect(renderedComponent.props().onElementDelete).toEqual(onElementDeleteSpy);
    expect(renderedComponent.props().onElementMove).toEqual(onElementSortSpy);
    elements.forEach((element, index) => {
      expect(renderedComponent.find(`ListElementClosable#le-${index}`).instance()).toBeInstanceOf(ListElementClosable);
      expect(renderedComponent.find(`ListElementClosable#le-${index}`).props().onElementNodeCreate).toEqual(renderedComponent.instance().handleElementNodeCreate);
      expect(renderedComponent.find(`li#le-${index}`).props().onMouseDown).toEqual(renderedComponent.find(`ListElementClosable#le-${index}`).instance().handleMouseDown);
      expect(renderedComponent.find(`li#le-${index} > div > button`).props().onClick).toEqual(renderedComponent.instance().handleElementDelete);
      expect(renderedComponent.find(`li#le-${index} > div > span`).text()).toEqual(element.text);
    });
  });

  it('input actions are work fine', () => {
    // Указываем высоту элементов, координаты и бордеры вручную,
    // так как getComputedStyle в среде JSDOM возвращает пустые строки вместо значений
    const elementHeights = [
      '55.375px',
      '55.375px',
      '55.375px'
    ];
    elementHeights.forEach((height, index) => {
      renderedComponent.find('li#le-' + index).instance().style.top = '0px';
      renderedComponent.find('li#le-' + index).instance().style.height = height;
      renderedComponent.find('li#le-' + index).instance().style.border = '1px';
    });

    renderedComponent.find('li#le-0').simulate('mousedown', { clientY: 1 });
    simulateMouseMove(1, 'down', 120);
    dispatchEvent(window.document.documentElement, getCustomMouseEvent("mouseup"));
    expect(onElementSortSpy).toBeCalledWith(0, 2);

    renderedComponent.find('li#le-1 > div > button').simulate('click');
    expect(onElementDeleteSpy).toBeCalledWith(2);
  });

  it('the snapshot and the rendered component are the same', () => {
    expect(toJson(renderedComponent)).toMatchSnapshot();
  });
});
