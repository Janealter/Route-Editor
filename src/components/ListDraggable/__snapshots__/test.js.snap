// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ListDraggable the snapshot and the rendered component are the same 1`] = `
<ListDraggable
  elements={
    Array [
      Object {
        "isRemovable": true,
        "key": 1,
        "text": "element1",
      },
      Object {
        "isRemovable": true,
        "key": 2,
        "text": "element2",
      },
      Object {
        "isRemovable": true,
        "key": 3,
        "text": "element3",
      },
    ]
  }
  listElementTemplate={[Function]}
  onElementDelete={
    [MockFunction] {
      "calls": Array [
        Array [
          2,
        ],
      ],
    }
  }
  onElementSort={
    [MockFunction] {
      "calls": Array [
        Array [
          0,
          2,
        ],
      ],
    }
  }
>
  <ul
    className="list-group"
  >
    <ListElementClosable
      id="le-0"
      isClosable={true}
      isHighlightable={true}
      key="1"
      onCloseButtonClick={[Function]}
      onElementNodeCreate={[Function]}
      onMouseDown={[Function]}
      text="element1"
    >
      <li
        className="list-element list-group-item"
        id="le-0"
        onMouseDown={[Function]}
      >
        <div
          className="list-element__inner"
        >
          <span
            className="list-element__text"
          >
            element1
          </span>
          <button
            className="list-element__button btn btn-danger btn-sm float-right"
            id="le-0"
            onClick={[Function]}
          >
            x
          </button>
        </div>
      </li>
    </ListElementClosable>
    <ListElementClosable
      id="le-1"
      isClosable={true}
      isHighlightable={true}
      key="2"
      onCloseButtonClick={[Function]}
      onElementNodeCreate={[Function]}
      onMouseDown={[Function]}
      text="element2"
    >
      <li
        className="list-element list-group-item"
        id="le-1"
        onMouseDown={[Function]}
      >
        <div
          className="list-element__inner"
        >
          <span
            className="list-element__text"
          >
            element2
          </span>
          <button
            className="list-element__button btn btn-danger btn-sm float-right"
            id="le-1"
            onClick={[Function]}
          >
            x
          </button>
        </div>
      </li>
    </ListElementClosable>
    <ListElementClosable
      id="le-2"
      isClosable={true}
      isHighlightable={true}
      key="3"
      onCloseButtonClick={[Function]}
      onElementNodeCreate={[Function]}
      onMouseDown={[Function]}
      text="element3"
    >
      <li
        className="list-element list-group-item"
        id="le-2"
        onMouseDown={[Function]}
      >
        <div
          className="list-element__inner"
        >
          <span
            className="list-element__text"
          >
            element3
          </span>
          <button
            className="list-element__button btn btn-danger btn-sm float-right"
            id="le-2"
            onClick={[Function]}
          >
            x
          </button>
        </div>
      </li>
    </ListElementClosable>
  </ul>
</ListDraggable>
`;
